ambry-replication/src/main/java/com.github.ambry.replication/BlobIdTransformer.java

```diff
@@ -158,10 +158,14 @@ private Message newMessage(InputStream inputStream, StoreKey newKey, MessageInfo
       PutMessageFormatInputStream putMessageFormatInputStream =
           new PutMessageFormatInputStream(newKey, blobEncryptionKey, newProperties, userMetaData, blobData.getStream(),
               blobData.getSize(), blobData.getBlobType());
+      // Reuse the CRC if present in the oldMessageInfo. This is important to ensure that messages that are received via
+      // replication are sent to the store with proper CRCs (which the store needs to detect duplicate messages).
+      // As an additional guard, here the CRC is only reused if the key's ID in string form is the same after conversion.
+      Long crc = oldMessageInfo.getStoreKey().getID().equals(newKey.getID()) ? oldMessageInfo.getCrc() : null;
```

Comment:
Just to be clear, I think this is the right change but the naming is the only thing that may be wrong. We can even drop the key equality check if we name this differently (`duplicateDetectionHelper`?)

Reply:
The crc here refers to the crc generated by the frontend for the blob when it was originally put. If we stick to that meaning the variable is probably correctly named. Plus, just changing the variable name may not be sufficient here since the method names and arguments for the constructor of `MessageInfo` all refer to crcs anyway.

Perhaps I can rename this as `originalCrc`. What do you think?